File: unity_app/Assets/Scripts/Data/PhoneAppData.cs
using System;
using System.Collections.Generic;
using EventPlannerSim.Core;

namespace EventPlannerSim.Data
{
    /// <summary>
    /// Data structure for Calendar app entries.
    /// Requirements: R4.2
    /// </summary>
    [Serializable]
    public class CalendarEntry
    {
        public string id;
        public string title;
        public string description;
        public GameDate date;
        public CalendarEntryType entryType;
        public string relatedEventId;
        public string relatedTaskId;
        public bool isCompleted;
        public bool isUrgent;
        
        /// <summary>
        /// Check if this entry is for today or past due.
        /// </summary>
        public bool IsDueOrPast(GameDate currentDate)
        {
            return date.TotalDays <= currentDate.TotalDays;
        }
        
        /// <summary>
        /// Get days until this entry's date.
        /// </summary>
        public int DaysUntil(GameDate currentDate)
        {
            return GameDate.DaysBetween(currentDate, date);
        }
    }
    
    /// <summary>
    /// Types of calendar entries.
    /// </summary>
    public enum CalendarEntryType
    {
        Event,          // Scheduled event
        TaskDeadline,   // Task deadline
        PrepTask,       // Preparation task
        Meeting,        // Client/vendor meeting
        Reminder        // General reminder
    }
    
    /// <summary>
    /// Data structure for Messages app entries.
    /// Requirements: R4.3, R4.6, R4.12, R4.13
    /// </summary>
    [Serializable]
    public class MessageThread
    {
        public string threadId;
        public string contactName;
        public string contactId;
        public MessageContactType contactType;
        public List<Message> messages = new();
        public bool hasUnread;
        public GameDate lastMessageDate;
        public bool isCompanyRelated; // Stage 2: company vs personal
        
        /// <summary>
        /// Get the most recent message in the thread.
        /// </summary>
        public Message GetLatestMessage()
        {
            if (messages == null || messages.Count == 0)
                return null;
            return messages[messages.Count - 1];
        }
        
        /// <summary>
        /// Get count of unread messages.
        /// </summary>
        public int GetUnreadCount()
        {
            int count = 0;
            foreach (var msg in messages)
            {
                if (!msg.isRead) count++;
            }
            return count;
        }
    }
    
    /// <summary>
    /// Individual message within a thread.
    /// </summary>
    [Serializable]
    public class Message
    {
        public string messageId;
        public string content;
        public GameDate sentDate;
        public bool isFromPlayer;
        public bool isRead;
        public MessagePriority priority;
    }
    
    /// <summary>
    /// Types of message contacts.
    /// </summary>
    public enum MessageContactType
    {
        Client,
        Vendor,
        Company,    // Stage 2: employer
        System      // Game notifications
    }
    
    /// <summary>
    /// Message priority levels.
    /// </summary>
    public enum MessagePriority
    {
        Normal,
        Important,
        Urgent
    }
    
    /// <summary>
    /// Data structure for Bank app entries.
    /// Requirements: R4.4, R4.14
    /// </summary>
    [Serializable]
    public class BankTransaction
    {
        public string transactionId;
        public string description;
        public float amount;
        public GameDate date;
        public TransactionType transactionType;
        public TransactionCategory category;
        public string relatedEventId;
        public bool isPending;
        public bool isCompanyRelated; // Stage 2: company vs side gig
    }
    
    /// <summary>
    /// Types of bank transactions.
    /// </summary>
    public enum TransactionType
    {
        Income,
        Expense,
        Pending
    }
    
    /// <summary>
    /// Categories for bank transactions.
    /// </summary>
    public enum TransactionCategory
    {
        EventPayment,       // Payment from client
        VendorPayment,      // Payment to vendor
        VenuePayment,       // Payment to venue
        Salary,             // Stage 2: company salary
        Commission,         // Stage 2: company commission
        SideGigIncome,      // Stage 2: personal side gig income
        FamilyHelp,         // Emergency family funding
        Contingency,        // Contingency fund usage
        Other
    }
    
    /// <summary>
    /// Bank account summary data.
    /// </summary>
    [Serializable]
    public class BankAccountSummary
    {
        public float currentBalance;
        public float pendingIncome;
        public float pendingExpenses;
        
        // Stage 2 separation
        public float salaryEarnings;
        public float commissionEarnings;
        public float sideGigEarnings;
        
        /// <summary>
        /// Get total available funds (balance + pending income - pending expenses).
        /// </summary>
        public float GetAvailableFunds()
        {
            return currentBalance + pendingIncome - pendingExpenses;
        }
    }
    
    /// <summary>
    /// Data structure for Contacts app entries (vendor rolodex).
    /// Requirements: R4.5
    /// </summary>
    [Serializable]
    public class ContactEntry
    {
        public string contactId;
        public string name;
        public ContactType contactType;
        public VendorCategory vendorCategory; // Only for vendor contacts
        public VendorTier vendorTier;         // Only for vendor contacts
        public float rating;
        public string playerNotes;
        public int relationshipLevel;
        public bool reliabilityRevealed;
        public bool flexibilityRevealed;
        public int timesHired;
        public GameDate lastHiredDate;
        public bool isFavorite;
    }
    
    /// <summary>
    /// Types of contacts.
    /// </summary>
    public enum ContactType
    {
        Vendor,
        Client,
        Venue,
        Company     // Stage 2: employer contacts
    }
    
    /// <summary>
    /// Data structure for Reviews app entries.
    /// Requirements: R4.9
    /// </summary>
    [Serializable]
    public class ReviewEntry
    {
        public string reviewId;
        public string clientName;
        public string eventTitle;
        public string eventId;
        public GameDate eventDate;
        public float satisfactionScore;
        public string testimonialText;
        public bool isPublic;
        public bool isHighlighted;
    }
    
    /// <summary>
    /// Reputation summary for Reviews app.
    /// </summary>
    [Serializable]
    public class ReputationSummary
    {
        public int currentReputation;
        public int totalEventsCompleted;
        public float averageSatisfaction;
        public int excellentEvents;     // 90%+ satisfaction
        public int goodEvents;          // 75-89% satisfaction
        public int averageEvents;       // 50-74% satisfaction
        public int poorEvents;          // <50% satisfaction
        public int currentExcellenceStreak;
        public int bestExcellenceStreak;
    }
    
    /// <summary>
    /// Data structure for Tasks app entries.
    /// Requirements: R4.10
    /// </summary>
    [Serializable]
    public class TaskListEntry
    {
        public string eventId;
        public string eventTitle;
        public GameDate eventDate;
        public List<TaskItem> tasks = new();
        public int completedCount;
        public int totalCount;
        public int pendingCount;
        public int failedCount;
        
        /// <summary>
        /// Get completion percentage.
        /// </summary>
        public float GetCompletionPercent()
        {
            if (totalCount == 0) return 0f;
            return (float)completedCount / totalCount * 100f;
        }
        
        /// <summary>
        /// Check if all tasks are complete.
        /// </summary>
        public bool IsComplete()
        {
            return completedCount == totalCount && totalCount > 0;
        }
    }
    
    /// <summary>
    /// Individual task item for Tasks app.
    /// </summary>
    [Serializable]
    public class TaskItem
    {
        public string taskId;
        public string taskName;
        public string description;
        public TaskStatus status;
        public GameDate deadline;
        public int hoursRequired;
        public bool isOverdue;
        public bool usedCompanyHelp;
    }
    
    /// <summary>
    /// Data structure for Clients app entries.
    /// Requirements: R4.11
    /// </summary>
    [Serializable]
    public class ClientRecord
    {
        public string clientId;
        public string clientName;
        public ClientPersonality personality;
        public List<ClientEventHistory> eventHistory = new();
        public int totalEventsPlanned;
        public float averageSatisfaction;
        public bool isReferralSource;
        public int referralsGenerated;
        public string playerNotes;
        public GameDate firstEventDate;
        public GameDate lastEventDate;
    }
    
    /// <summary>
    /// Event history entry for a client.
    /// </summary>
    [Serializable]
    public class ClientEventHistory
    {
        public string eventId;
        public string eventTitle;
        public string eventTypeId; // References EventTypeData.eventTypeId
        public GameDate eventDate;
        public float satisfactionScore;
        public float budgetTotal;
        public bool wasReferral;
    }
}

File: EventApp/EventsApp/Models/Data/PhoneAppData.swift
import Foundation

// MARK: - Calendar

struct CalendarEntry: Codable, Equatable, Identifiable {
    var id: String
    var title: String
    var description: String
    var date: GameDate
    var entryType: CalendarEntryType
    var relatedEventId: String?
    var relatedTaskId: String?
    var isCompleted: Bool = false
    var isUrgent: Bool = false

    func isDueOrPast(currentDate: GameDate) -> Bool {
        currentDate >= date
    }

    func daysUntil(currentDate: GameDate) -> Int {
        currentDate.daysBetween(date)
    }
}

// MARK: - Messages

struct MessageThread: Codable, Equatable, Identifiable {
    var id: String { threadId }
    var threadId: String
    var contactName: String
    var contactId: String
    var contactType: MessageContactType
    var messages: [Message] = []
    var hasUnread: Bool = false
    var lastMessageDate: GameDate
    var isCompanyRelated: Bool = false

    func getLatestMessage() -> Message? {
        messages.last
    }

    func getUnreadCount() -> Int {
        messages.filter { !$0.isRead && !$0.isFromPlayer }.count
    }
}

struct Message: Codable, Equatable, Identifiable {
    var id: String { messageId }
    var messageId: String
    var content: String
    var sentDate: GameDate
    var isFromPlayer: Bool
    var isRead: Bool = false
    var priority: MessagePriority = .normal
}

// MARK: - Bank

struct BankTransaction: Codable, Equatable, Identifiable {
    var id: String { transactionId }
    var transactionId: String
    var description: String
    var amount: Double
    var date: GameDate
    var transactionType: TransactionType
    var category: TransactionCategory
    var relatedEventId: String?
    var isPending: Bool = false
    var isCompanyRelated: Bool = false
}

struct BankAccountSummary: Codable, Equatable {
    var currentBalance: Double = 0
    var pendingIncome: Double = 0
    var pendingExpenses: Double = 0
    var salaryEarnings: Double = 0
    var commissionEarnings: Double = 0
    var sideGigEarnings: Double = 0

    func getAvailableFunds() -> Double {
        currentBalance + pendingIncome - pendingExpenses
    }
}

// MARK: - Contacts

struct ContactEntry: Codable, Equatable, Identifiable {
    var id: String { contactId }
    var contactId: String
    var name: String
    var contactType: ContactType
    var vendorCategory: VendorCategory?
    var vendorTier: VendorTier?
    var rating: Double = 0
    var playerNotes: String = ""
    var relationshipLevel: Int = 0
    var reliabilityRevealed: Bool = false
    var flexibilityRevealed: Bool = false
    var timesHired: Int = 0
    var lastHiredDate: GameDate?
    var isFavorite: Bool = false
}

// MARK: - Reviews

struct ReviewEntry: Codable, Equatable, Identifiable {
    var id: String { reviewId }
    var reviewId: String
    var clientName: String
    var eventTitle: String
    var eventId: String
    var eventDate: GameDate
    var satisfactionScore: Double
    var testimonialText: String
    var isPublic: Bool = true
    var isHighlighted: Bool = false
}

struct ReputationSummary: Codable, Equatable {
    var currentReputation: Int = 0
    var totalEventsCompleted: Int = 0
    var averageSatisfaction: Double = 0
    var excellentEvents: Int = 0
    var goodEvents: Int = 0
    var averageEvents: Int = 0
    var poorEvents: Int = 0
    var currentExcellenceStreak: Int = 0
    var bestExcellenceStreak: Int = 0
}

// MARK: - Tasks

struct TaskListEntry: Codable, Equatable {
    var eventId: String
    var eventTitle: String
    var eventDate: GameDate
    var tasks: [TaskItem] = []
    var completedCount: Int = 0
    var totalCount: Int = 0
    var pendingCount: Int = 0
    var failedCount: Int = 0

    func getCompletionPercent() -> Double {
        guard totalCount > 0 else { return 0 }
        return Double(completedCount) / Double(totalCount) * 100.0
    }

    func isComplete() -> Bool {
        completedCount == totalCount && totalCount > 0
    }
}

struct TaskItem: Codable, Equatable {
    var taskId: String
    var taskName: String
    var description: String
    var status: TaskStatus
    var deadline: GameDate
    var hoursRequired: Int
    var isOverdue: Bool = false
    var usedCompanyHelp: Bool = false
}

// MARK: - Clients

struct ClientRecord: Codable, Equatable, Identifiable {
    var id: String { clientId }
    var clientId: String
    var clientName: String
    var personality: ClientPersonality
    var eventHistory: [ClientEventHistory] = []
    var totalEventsPlanned: Int = 0
    var averageSatisfaction: Double = 0
    var isReferralSource: Bool = false
    var referralsGenerated: Int = 0
    var playerNotes: String = ""
    var firstEventDate: GameDate?
    var lastEventDate: GameDate?
}

struct ClientEventHistory: Codable, Equatable {
    var eventId: String
    var eventTitle: String
    var eventTypeId: String
    var eventDate: GameDate
    var satisfactionScore: Double
    var budgetTotal: Double
    var wasReferral: Bool
}
